backend/postgres/postgres.go

```diff
@@ -0,0 +1,955 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	_ "embed"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/microsoft/durabletask-go/api"
+	"github.com/microsoft/durabletask-go/backend"
+	"github.com/microsoft/durabletask-go/internal/helpers"
+	"github.com/microsoft/durabletask-go/internal/protos"
+
+	"google.golang.org/protobuf/encoding/protojson"
+
+	_ "github.com/jackc/pgx/v5/stdlib"
+)
+
+//go:embed schema.sql
+var schema string
+
+var emptyString string = ""
+
+type postgresBackend struct {
+	connStr     string
+	db          *sql.DB
+	workerName  string
+	logger      backend.Logger
+	initLock    sync.Mutex
+	initialized bool
+
+	orchestrationLockTimeout time.Duration
+	activityLockTimeout      time.Duration
+}
+
+// NewPostgresBackend creates a new postgres-based Backend object.
+func NewPostgresBackend(connString string, logger backend.Logger) backend.Backend {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "unknown"
+	}
+
+	pid := os.Getpid()
+	uuidStr := uuid.NewString()
+
+	be := &postgresBackend{
+		connStr:                  connString,
+		workerName:               fmt.Sprintf("%s,%d,%s", hostname, pid, uuidStr),
+		logger:                   logger,
+		orchestrationLockTimeout: 2 * time.Minute,
+		activityLockTimeout:      2 * time.Minute,
+	}
+
+	return be
+}
+
+// CreateTaskHub creates the postgres database and applies the schema
+func (be *postgresBackend) CreateTaskHub(ctx context.Context) error {
+	be.initLock.Lock()
+	defer be.initLock.Unlock()
+
+	if be.db == nil {
+		be.logger.Debug("Opening database connection")
+		db, err := sql.Open("pgx", be.connStr)
+		if err != nil {
+			return fmt.Errorf("failed to open the database: %w", err)
+		}
+		be.db = db
+	}
+
+	if !be.initialized {
+		// Initialize database
+		be.logger.Info("Running database initialization script")
```

Comment:
How will you handle database schema migrations?

Reply:
When we get around to supporting database schema migrations, the path I typically follow is having versioned schema scripts: a base script and then delta scripts for all schema changes (adding columns, indexes, etc.).

Reply:
Yes, that sounds reasonable.

I would suggest considering adding the "migrations" table from day 0 however. Eventually, migrations will need to be done. We can also import the library we use in Dapr, which supports database/sql and pgx already: https://github.com/dapr/components-contrib/blob/master/internal/component/sql/migrations.go